# 01 — Data Understanding

## Project: Avenix — Lead Scoring ML

### Objective

Understand the structure, quality, and characteristics of the lead dataset before performing data cleaning and feature engineering.

### We will analyze:

- Dataset shape
- Feature names
- Data types
- Target variable
- Missing values
- Duplicate records
- Unique values
- Numerical and categorical features
- Basic statistics

In [28]:
import pandas as pd
import numpy as np

In [29]:
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [30]:
DATA_PATH = "../data/raw/Leads X Education.csv"
df = pd.read_csv(DATA_PATH)

## 1. Dataset Overview

First, we inspect the first and last rows to understand what the dataset looks like.

In [31]:
df.head()

,Prospect ID,Lead Number,Lead Origin,Lead Source,Do Not Email,Do Not Call,Converted,TotalVisits,Total Time Spent on Website,Page Views Per Visit,Last Activity,Country,Specialization,How did you hear about X Education,What is your current occupation,What matters most to you in choosing a course,Search,Magazine,Newspaper Article,X Education Forums,Newspaper,Digital Advertisement,Through Recommendations,Receive More Updates About Our Courses,Tags,Lead Quality,Update me on Supply Chain Content,Get updates on DM Content,Lead Profile,City,Asymmetrique Activity Index,Asymmetrique Profile Index,Asymmetrique Activity Score,Asymmetrique Profile Score,I agree to pay the amount through cheque,A free copy of Mastering The Interview,Last Notable Activity
0,7927b2df-8bba-4d29-b9a2-b6e0beafe620,660737,API,Olark Chat,No,No,0,0.0,0,0.0,Page Visited on Website,NaN,Select,Select,Unemployed,Better Career Prospects,No,No,No,No,No,No,No,No,Interested in other courses,Low in Relevance,No,No,Select,Select,02.Medium,02.Medium,15.0,15.0,No,No,Modified
1,2a272436-5132-4136-86fa-dcc88c88f482,660728,API,Organic Search,No,No,0,5.0,674,2.5,Email Opened,India,Select,Select,Unemployed,Better Career Prospects,No,No,No,No,No,No,No,No,Ringing,NaN,No,No,Select,Select,02.Medium,02.Medium,15.0,15.0,No,No,Email Opened
2,8cc8c611-a219-4f35-ad23-fdfd2656bd8a,660727,Landing Page Submission,Direct Traffic,No,No,1,2.0,1532,2.0,Email Opened,India,Business Administration,Select,Student,Better Career Prospects,No,No,No,No,No,No,No,No,Will revert after reading the email,Might be,No,No,Potential Lead,Mumbai,02.Medium,01.High,14.0,20.0,No,Yes,Email Opened
3,0cc2df48-7cf4-4e39-9de9-19797f9b38cc,660719,Landing Page Submission,Direct Traffic,No,No,0,1.0,305,1.0,Unreachable,India,Media and Advertising,Word Of Mouth,Unemployed,Better Career Prospects,No,No,No,No,No,No,No,No,Ringing,Not Sure,No,No,Select,Mumbai,02.Medium,01.High,13.0,17.0,No,No,Modified
4,3256f628-e534-4826-9d63-4a8b88782852,660681,Landing Page Submission,Google,No,No,1,2.0,1428,1.0,Converted to Lead,India,Select,Other,Unemployed,Better Career Prospects,No,No,No,No,No,No,No,No,Will revert after reading the email,Might be,No,No,Select,Mumbai,02.Medium,01.High,15.0,18.0,No,No,Modified


In [32]:
df.tail()

,Prospect ID,Lead Number,Lead Origin,Lead Source,Do Not Email,Do Not Call,Converted,TotalVisits,Total Time Spent on Website,Page Views Per Visit,Last Activity,Country,Specialization,How did you hear about X Education,What is your current occupation,What matters most to you in choosing a course,Search,Magazine,Newspaper Article,X Education Forums,Newspaper,Digital Advertisement,Through Recommendations,Receive More Updates About Our Courses,Tags,Lead Quality,Update me on Supply Chain Content,Get updates on DM Content,Lead Profile,City,Asymmetrique Activity Index,Asymmetrique Profile Index,Asymmetrique Activity Score,Asymmetrique Profile Score,I agree to pay the amount through cheque,A free copy of Mastering The Interview,Last Notable Activity
9235,19d6451e-fcd6-407c-b83b-48e1af805ea9,579564,Landing Page Submission,Direct Traffic,Yes,No,1,8.0,1845,2.67,Email Marked Spam,Saudi Arabia,IT Projects Management,Select,Unemployed,Better Career Prospects,No,No,No,No,No,No,No,No,Will revert after reading the email,High in Relevance,No,No,Potential Lead,Mumbai,02.Medium,01.High,15.0,17.0,No,No,Email Marked Spam
9236,82a7005b-7196-4d56-95ce-a79f937a158d,579546,Landing Page Submission,Direct Traffic,No,No,0,2.0,238,2.00,SMS Sent,India,Media and Advertising,Select,Unemployed,Better Career Prospects,No,No,No,No,No,No,No,No,wrong number given,Might be,No,No,Potential Lead,Mumbai,02.Medium,01.High,14.0,19.0,No,Yes,SMS Sent
9237,aac550fe-a586-452d-8d3c-f1b62c94e02c,579545,Landing Page Submission,Direct Traffic,Yes,No,0,2.0,199,2.00,SMS Sent,India,Business Administration,Select,Unemployed,Better Career Prospects,No,No,No,No,No,No,No,No,invalid number,Not Sure,No,No,Potential Lead,Mumbai,02.Medium,01.High,13.0,20.0,No,Yes,SMS Sent
9238,5330a7d1-2f2b-4df4-85d6-64ca2f6b95b9,579538,Landing Page Submission,Google,No,No,1,3.0,499,3.00,SMS Sent,India,Human Resource Management,Online Search,NaN,NaN,No,No,No,No,No,No,No,No,NaN,NaN,No,No,NaN,Other Metro Cities,02.Medium,02.Medium,15.0,16.0,No,No,SMS Sent
9239,571b5c8e-a5b2-4d57-8574-f2ffb06fdeff,579533,Landing Page Submission,Direct Traffic,No,No,1,6.0,1279,3.00,SMS Sent,Bangladesh,Supply Chain Management,Select,Unemployed,Better Career Prospects,No,No,No,No,No,No,No,No,Will revert after reading the email,Might be,No,No,Potential Lead,Other Cities,02.Medium,01.High,15.0,18.0,No,Yes,Modified


## 2. Dataset Shape

We check the number of rows and columns in the dataset.


In [33]:
df.shape

(9240, 37)

## 3. Column Names

We inspect all available features and identify the target variable.

In [34]:
for i, column in enumerate(df.columns, start=1):
    print(f"{i}. {column}")

1. Prospect ID
2. Lead Number
3. Lead Origin
4. Lead Source
5. Do Not Email
6. Do Not Call
7. Converted
8. TotalVisits
9. Total Time Spent on Website
10. Page Views Per Visit
11. Last Activity
12. Country
13. Specialization
14. How did you hear about X Education
15. What is your current occupation
16. What matters most to you in choosing a course
17. Search
18. Magazine
19. Newspaper Article
20. X Education Forums
21. Newspaper
22. Digital Advertisement
23. Through Recommendations
24. Receive More Updates About Our Courses
25. Tags
26. Lead Quality
27. Update me on Supply Chain Content
28. Get updates on DM Content
29. Lead Profile
30. City
31. Asymmetrique Activity Index
32. Asymmetrique Profile Index
33. Asymmetrique Activity Score
34. Asymmetrique Profile Score
35. I agree to pay the amount through cheque
36. A free copy of Mastering The Interview
37. Last Notable Activity


## 4. Data Types

Understanding data types helps us distinguish numerical and categorical features.

In [35]:
df.dtypes

Prospect ID                                          str
Lead Number                                        int64
Lead Origin                                          str
Lead Source                                          str
Do Not Email                                         str
Do Not Call                                          str
Converted                                          int64
TotalVisits                                      float64
Total Time Spent on Website                        int64
Page Views Per Visit                             float64
Last Activity                                        str
Country                                              str
Specialization                                       str
How did you hear about X Education                   str
What is your current occupation                      str
What matters most to you in choosing a course        str
Search                                               str
Magazine                       

In [36]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 37 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Prospect ID                                    9240 non-null   str    
 1   Lead Number                                    9240 non-null   int64  
 2   Lead Origin                                    9240 non-null   str    
 3   Lead Source                                    9204 non-null   str    
 4   Do Not Email                                   9240 non-null   str    
 5   Do Not Call                                    9240 non-null   str    
 6   Converted                                      9240 non-null   int64  
 7   TotalVisits                                    9103 non-null   float64
 8   Total Time Spent on Website                    9240 non-null   int64  
 9   Page Views Per Visit                           9103 non-null   

In [37]:
numeric_columns = df.select_dtypes(
    include=["int64", "float64"]
).columns

categorical_columns = df.select_dtypes(
    include=["object", "category"]
).columns

/tmp/ipykernel_45075/3072391946.py:5: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(


In [38]:
print("Numerical columns:")
print(numeric_columns.tolist())

print("\nCategorical columns:")
print(categorical_columns.tolist())

Numerical columns:
['Lead Number', 'Converted', 'TotalVisits', 'Total Time Spent on Website', 'Page Views Per Visit', 'Asymmetrique Activity Score', 'Asymmetrique Profile Score']

Categorical columns:
['Prospect ID', 'Lead Origin', 'Lead Source', 'Do Not Email', 'Do Not Call', 'Last Activity', 'Country', 'Specialization', 'How did you hear about X Education', 'What is your current occupation', 'What matters most to you in choosing a course', 'Search', 'Magazine', 'Newspaper Article', 'X Education Forums', 'Newspaper', 'Digital Advertisement', 'Through Recommendations', 'Receive More Updates About Our Courses', 'Tags', 'Lead Quality', 'Update me on Supply Chain Content', 'Get updates on DM Content', 'Lead Profile', 'City', 'Asymmetrique Activity Index', 'Asymmetrique Profile Index', 'I agree to pay the amount through cheque', 'A free copy of Mastering The Interview', 'Last Notable Activity']


## 5. Missing Values

We identify columns containing missing values and measure their amount.

Missing values must be analyzed before deciding how they should be handled.

In [39]:
missing_count = df.isnull().sum()

missing_count = missing_count[missing_count > 0]

missing_count.sort_values(ascending=False)

Lead Quality                                     4767
Asymmetrique Profile Score                       4218
Asymmetrique Activity Index                      4218
Asymmetrique Activity Score                      4218
Asymmetrique Profile Index                       4218
Tags                                             3353
What matters most to you in choosing a course    2709
Lead Profile                                     2709
What is your current occupation                  2690
Country                                          2461
How did you hear about X Education               2207
Specialization                                   1438
City                                             1420
Page Views Per Visit                              137
TotalVisits                                       137
Last Activity                                     103
Lead Source                                        36
dtype: int64

In [40]:
missing_percentage = (
    df.isnull().mean() * 100
).sort_values(ascending=False)

missing_percentage[missing_percentage > 0]

Lead Quality                                     51.590909
Asymmetrique Profile Index                       45.649351
Asymmetrique Activity Score                      45.649351
Asymmetrique Activity Index                      45.649351
Asymmetrique Profile Score                       45.649351
Tags                                             36.287879
What matters most to you in choosing a course    29.318182
Lead Profile                                     29.318182
What is your current occupation                  29.112554
Country                                          26.634199
How did you hear about X Education               23.885281
Specialization                                   15.562771
City                                             15.367965
Page Views Per Visit                              1.482684
TotalVisits                                       1.482684
Last Activity                                     1.114719
Lead Source                                       0.3896

In [41]:
missing_df = pd.DataFrame({
    "missing_count": df.isnull().sum(),
    "missing_percentage": df.isnull().mean() * 100
})

missing_df = missing_df[
    missing_df["missing_count"] > 0
].sort_values(
    "missing_percentage",
    ascending=False
)

missing_df

,missing_count,missing_percentage
Lead Quality,4767,51.590909
Asymmetrique Profile Score,4218,45.649351
Asymmetrique Activity Index,4218,45.649351
Asymmetrique Activity Score,4218,45.649351
Asymmetrique Profile Index,4218,45.649351
Tags,3353,36.287879
What matters most to you in choosing a course,2709,29.318182
Lead Profile,2709,29.318182
What is your current occupation,2690,29.112554
Country,2461,26.634199


## 6. Duplicate Records

We check whether the dataset contains duplicated rows.

In [42]:
duplicate_count = df.duplicated().sum()

print(f"Duplicate rows: {duplicate_count}")

Duplicate rows: 0


In [43]:
df[df.duplicated()].head()

,Prospect ID,Lead Number,Lead Origin,Lead Source,Do Not Email,Do Not Call,Converted,TotalVisits,Total Time Spent on Website,Page Views Per Visit,Last Activity,Country,Specialization,How did you hear about X Education,What is your current occupation,What matters most to you in choosing a course,Search,Magazine,Newspaper Article,X Education Forums,Newspaper,Digital Advertisement,Through Recommendations,Receive More Updates About Our Courses,Tags,Lead Quality,Update me on Supply Chain Content,Get updates on DM Content,Lead Profile,City,Asymmetrique Activity Index,Asymmetrique Profile Index,Asymmetrique Activity Score,Asymmetrique Profile Score,I agree to pay the amount through cheque,A free copy of Mastering The Interview,Last Notable Activity


In [44]:
for column in categorical_columns:
    print(f"\n{'=' * 60}")
    print(column)
    print(f"Unique values: {df[column].nunique()}")
    print(df[column].value_counts(dropna=False).head(10))


Prospect ID
Unique values: 9240
Prospect ID
7927b2df-8bba-4d29-b9a2-b6e0beafe620    1
2a272436-5132-4136-86fa-dcc88c88f482    1
8cc8c611-a219-4f35-ad23-fdfd2656bd8a    1
0cc2df48-7cf4-4e39-9de9-19797f9b38cc    1
3256f628-e534-4826-9d63-4a8b88782852    1
2058ef08-2858-443e-a01f-a9237db2f5ce    1
9fae7df4-169d-489b-afe4-0f3d752542ed    1
20ef72a2-fb3b-45e0-924e-551c5fa59095    1
cfa0128c-a0da-4656-9d47-0aa4e67bf690    1
af465dfc-7204-4130-9e05-33231863c4b5    1
Name: count, dtype: int64

Lead Origin
Unique values: 5
Lead Origin
Landing Page Submission    4886
API                        3580
Lead Add Form               718
Lead Import                  55
Quick Add Form                1
Name: count, dtype: int64

Lead Source
Unique values: 21
Lead Source
Google              2868
Direct Traffic      2543
Olark Chat          1755
Organic Search      1154
Reference            534
Welingak Website     142
Referral Sites       125
Facebook              55
NaN                   36
bing         

## 7. Numerical Feature Statistics

We inspect descriptive statistics for numerical variables.

In [45]:
df[numeric_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
Lead Number,9240.0,617188.435606,23405.995698,579533.0,596484.5,615479.0,637387.25,660737.0
Converted,9240.0,0.385390,0.486714,0.0,0.0,0.0,1.00,1.0
TotalVisits,9103.0,3.445238,4.854853,0.0,1.0,3.0,5.00,251.0
Total Time Spent on Website,9240.0,487.698268,548.021466,0.0,12.0,248.0,936.00,2272.0
Page Views Per Visit,9103.0,2.362820,2.161418,0.0,1.0,2.0,3.00,55.0
Asymmetrique Activity Score,5022.0,14.306252,1.386694,7.0,14.0,14.0,15.00,18.0
Asymmetrique Profile Score,5022.0,16.344883,1.811395,11.0,15.0,16.0,18.00,20.0


In [46]:
TARGET = "Converted"
print(df[TARGET].value_counts())

Converted
0    5679
1    3561
Name: count, dtype: int64


In [47]:
print(df[TARGET].value_counts(normalize=True) * 100)

Converted
0    61.461039
1    38.538961
Name: proportion, dtype: float64


## 8. Target Variable

The target variable represents whether a lead converted.

We inspect the class distribution to understand whether the target is balanced or imbalanced.

In [48]:
target_distribution = df[TARGET].value_counts().to_frame("count")

target_distribution["percentage"] = (
    target_distribution["count"]
    / len(df)
    * 100
)

target_distribution

,count,percentage
Converted,,
0,5679,61.461039
1,3561,38.538961


In [49]:
print("=== MISSING VALUES ===")
print(missing_df)

print("\n=== DUPLICATES ===")
print(df.duplicated().sum())

print("\n=== TARGET COUNT ===")
print(df["Converted"].value_counts())

print("\n=== TARGET PERCENTAGE ===")
print(df["Converted"].value_counts(normalize=True) * 100)

=== MISSING VALUES ===
                                               missing_count  \
Lead Quality                                            4767   
Asymmetrique Profile Score                              4218   
Asymmetrique Activity Index                             4218   
Asymmetrique Activity Score                             4218   
Asymmetrique Profile Index                              4218   
Tags                                                    3353   
What matters most to you in choosing a course           2709   
Lead Profile                                            2709   
What is your current occupation                         2690   
Country                                                 2461   
How did you hear about X Education                      2207   
Specialization                                          1438   
City                                                    1420   
Page Views Per Visit                                     137   
TotalVisits      

# Data Understanding Summary

### Dataset

- Number of rows: 9240
- Number of columns: 37

### Features

- Numerical features: 'Lead Number', 'Converted', 'TotalVisits', 'Total Time Spent on Website', 'Page Views Per Visit', 'Asymmetrique Activity Score', 'Asymmetrique Profile Score'
- Categorical features: 'Prospect ID', 'Lead Origin', 'Lead Source', 'Do Not Email', 'Do Not Call', 'Last Activity', 'Country', 'Specialization', 'How did you hear about X Education', 'What is your current occupation', 'What matters most to you in choosing a course', 'Search', 'Magazine', 'Newspaper Article', 'X Education Forums', 'Newspaper', 'Digital Advertisement', 'Through Recommendations', 'Receive More Updates About Our Courses', 'Tags', 'Lead Quality', 'Update me on Supply Chain Content', 'Get updates on DM Content', 'Lead Profile', 'City', 'Asymmetrique Activity Index', 'Asymmetrique Profile Index', 'I agree to pay the amount through cheque', 'A free copy of Mastering The Interview', 'Last Notable Activity'

### Data Quality

- Missing values: 
Lead Quality                                            4767   
Asymmetrique Profile Score                              4218   
Asymmetrique Activity Index                             4218   
Asymmetrique Activity Score                             4218   
Asymmetrique Profile Index                              4218   
Tags                                                    3353   
What matters most to you in choosing a course           2709   
Lead Profile                                            2709   
What is your current occupation                         2690   
Country                                                 2461   
How did you hear about X Education                      2207   
Specialization                                          1438   
City                                                    1420   
Page Views Per Visit                                     137   
TotalVisits                                              137   
Last Activity                                            103   
Lead Source                                               36 
Lead Quality                                            51.590909  
Asymmetrique Profile Score                              45.649351  
Asymmetrique Activity Index                             45.649351  
Asymmetrique Activity Score                             45.649351  
Asymmetrique Profile Index                              45.649351  
Tags                                                    36.287879  
What matters most to you in choosing a course           29.318182  
Lead Profile                                            29.318182  
What is your current occupation                         29.112554  
Country                                                 26.634199  
How did you hear about X Education                      23.885281  
Specialization                                          15.562771  
City                                                    15.367965  
Page Views Per Visit                                     1.482684  
TotalVisits                                              1.482684  
Last Activity                                            1.114719  
Lead Source                                              0.389610 
- Duplicate rows: 0

### Target

=== TARGET COUNT ===
Converted
0    5679
1    3561
Name: count, dtype: int64

=== TARGET PERCENTAGE ===
Converted
0    61.461039
1    38.538961
Name: proportion, dtype: float64

### Initial Observations

- Several categorical features require encoding.
- Missing values require further investigation.
- Numerical variables should be checked for outliers.
- Target distribution should be considered during model validation.
- Potential data leakage must be investigated before training.